src/bundle-manifest.ts

```diff
@@ -359,24 +425,33 @@ export function mergeBundles(bundles: Bundle[]): Bundle {
         new Set<ResolvedUrl>([...newBundle.inlinedScripts, ...inlinedScripts]);
     newBundle.inlinedStyles =
         new Set<ResolvedUrl>([...newBundle.inlinedStyles, ...inlinedStyles]);
+    newBundle.bundledExports = new Map<ResolvedUrl, Map<string, string>>(
+        [...newBundle.bundledExports, ...bundledExports]);
   }
   return newBundle;
 }
 
 /**
- * Return a new bundle array where all bundles within it matching the predicate
- * are merged.
+ * Return a new bundle array where bundles within it matching the predicate
+ * are merged together.  Note that merge operations are segregated by type so
+ * that no attempt to merge bundles of different types will occur.
  */
 export function mergeMatchingBundles(
     bundles: Bundle[], predicate: (bundle: Bundle) => boolean): Bundle[] {
   const newBundles = Array.from(bundles);
   const bundlesToMerge = newBundles.filter(predicate);
-  if (bundlesToMerge.length > 1) {
-    for (const bundle of bundlesToMerge) {
-      newBundles.splice(newBundles.indexOf(bundle), 1);
-    }
-    newBundles.push(mergeBundles(bundlesToMerge));
-  }
+
+  [...new Set(bundlesToMerge.map((b) => b.type))].sort().forEach(
+      (bundleType) => {
+        const bundlesToMergeForType =
+            bundlesToMerge.filter((b) => b.type === bundleType);
+        if (bundlesToMergeForType.length > 1) {
+          for (const bundle of bundlesToMergeForType) {
+            newBundles.splice(newBundles.indexOf(bundle), 1);
+          }
+          newBundles.push(mergeBundles(bundlesToMergeForType));
```

Comment:
Are we sure this isn't moving a dependency to after it's dependants?

Reply:
moot after refactoring the method in commit ac90b3c48a906183b49d6a5dcbb8265084cce269